File: day1/day1.cpp
#include<fstream>
#include<iostream>

using namespace std;

int main(int argc, char **argv) {
    // Open at end of file
    ifstream infile(argv[1], ios::in|ios::ate);
    // Get length of file 
    size_t length = infile.tellg();
    // Seek to beginning of file
    infile.seekg(0, ios::beg);

    // Read one byte at a time
    char mychar;
    int sum = 0, first, prev, cur;
    // Grab first digit (no modular arithmetic here)
    infile.read(&mychar, sizeof(char));
    first = mychar - 48;
    prev = first;

    // Part 1
    for(int i=1; i < length - 1; i++) {
        infile.read(&mychar, sizeof(char));
        cur = mychar - 48;
        if (cur == prev) {
            sum += cur;
        }
        prev = cur;
    }
    if (prev == first) {
        sum += first;
    }
    cout << "Part1: " << sum << endl;

    // Part 2
    int partner;
    sum = 0;
    int stride = (length - 1) / 2;
    for(int i=0; i < stride; i++) {
        // from beginning, seek to i
        infile.seekg(i, ios::beg);
        infile.read(&mychar, sizeof(char));
        cur = mychar - 48;
        // seek to partner
        infile.seekg(stride - 1, ios::cur);
        infile.read(&mychar, sizeof(char));
        partner = mychar - 48;
        if (cur == partner) {
            sum += 2*cur;
        }
    }
    cout << "Part2: " << sum << endl;

    infile.close();

    return 0;
}

File: day1/day1.py
# Inverse Captcha

import sys

# Part 1
with open(sys.argv[1], 'r') as myfile:
    data = myfile.read().replace('\n', '')
    sum = 0
    for i in range(0, len(data)):
        if data[i] == data[(i - 1) % len(data)]:
            sum += int(data[i])
    print sum

# Part 2
with open(sys.argv[1], 'r') as myfile:
    data = myfile.read().replace('\n', '')
    sum = 0
    num_elements = len(data)
    for i in range(0, num_elements):
        if data[i] == data[(i + (num_elements / 2)) % num_elements]:
            sum += int(data[i])
    print sum
